7.20.0 regression: TestEngine with ID 'cucumber' failed to discover tests

### 👓 What did you see?

Build of PR opened by renovate with new Cucumber version is failing: https://github.com/jhipster/jhipster-lite/pull/11057:
>  Error:  Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.5.0:test (default-test) on project jhlite: 
Error:  
Error:  Please refer to /home/runner/work/jhipster-lite/jhipster-lite/target/surefire-reports for the individual test results.
Error:  Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
Error: [ERROR] There was an error in the forked process
Error:  TestEngine with ID 'cucumber' failed to discover tests
Error:  org.apache.maven.surefire.booter.SurefireBooterForkException: There was an error in the forked process
Error:  TestEngine with ID 'cucumber' failed to discover tests
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:628)
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:285)
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:250)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1337)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1137)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:971)
Error:  	at com.gradle.maven.cache.extension.d.c.a(SourceFile:27)

### ✅ What did you expect to see?

Build running fine, as it was with 7.19.0

### 📦 Which tool/library version are you using?

cucumber-junit-platform-engine
cucumber-spring
maven-failsafe-plugin
gradle-develocity

### 🔬 How could we reproduce it?

_No response_

### 📚 Any additional context?

Our cucumbers tests are executed with maven-failsafe-plugin, but here the error happens with the execution of maven-surefire-plugin

```shell
$ cat jhipster/jhipster-lite/target/surefire-reports/2024-10-05T16-15-25_011-jvmRun1.dump
# Created at 2024-10-05T16:15:26.116
org.junit.platform.commons.JUnitException: TestEngine with ID 'cucumber' failed to discover tests
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:160)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:132)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:78)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:99)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:77)
	at org.junit.platform.launcher.core.DelegatingLauncher.discover(DelegatingLauncher.java:42)
	at org.apache.maven.surefire.junitplatform.LazyLauncher.discover(LazyLauncher.java:50)
	at org.apache.maven.surefire.junitplatform.TestPlanScannerFilter.accept(TestPlanScannerFilter.java:52)
	at org.apache.maven.surefire.api.util.DefaultScanResult.applyFilter(DefaultScanResult.java:87)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.scanClasspath(JUnitPlatformProvider.java:142)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:122)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
Caused by: java.util.ServiceConfigurationError: io.cucumber.core.eventbus.UuidGenerator: Provider io.cucumber.core.eventbus.IncrementingUuidGenerator could not be instantiated
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:586)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:813)
	at java.base/java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:729)
	at java.base/java.util.ServiceLoader$3.next(ServiceLoader.java:1403)
	at io.cucumber.core.runtime.UuidGeneratorServiceLoader.loadSingleUuidGeneratorOrDefault(UuidGeneratorServiceLoader.java:60)
	at io.cucumber.core.runtime.UuidGeneratorServiceLoader.loadUuidGenerator(UuidGeneratorServiceLoader.java:46)
	at io.cucumber.junit.platform.engine.FeatureResolver.createFeatureParser(FeatureResolver.java:69)
	at io.cucumber.junit.platform.engine.FeatureResolver.<init>(FeatureResolver.java:59)
	at io.cucumber.junit.platform.engine.FeatureResolver.create(FeatureResolver.java:78)
	at io.cucumber.junit.platform.engine.DiscoverySelectorResolver.resolve(DiscoverySelectorResolver.java:67)
	at io.cucumber.junit.platform.engine.DiscoverySelectorResolver.resolveSelectors(DiscoverySelectorResolver.java:53)
	at io.cucumber.junit.platform.engine.CucumberTestEngine.discover(CucumberTestEngine.java:47)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:152)
	... 15 more
Caused by: io.cucumber.core.exception.CucumberException: Out of IncrementingUuidGenerator capacity. Please reuse existing instances or use another UuidGenerator implementation instead.
	at io.cucumber.core.eventbus.IncrementingUuidGenerator.<init>(IncrementingUuidGenerator.java:102)
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:789)
	... 26 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Fixed
+- [Core] Lazily start IncrementingUuidGenerator sessions([#2931](https://github.com/cucumber/cucumber-jvm/pull/2931) M.P. Korstanje)
 
 ## [7.20.0] - 2024-10-04
 ### Added

--- a/cucumber-core/src/main/java/io/cucumber/core/eventbus/IncrementingUuidGenerator.java
+++ b/cucumber-core/src/main/java/io/cucumber/core/eventbus/IncrementingUuidGenerator.java
@@ -68,7 +68,7 @@ public class IncrementingUuidGenerator implements UuidGenerator {
     /**
      * Computed UUID MSB value.
      */
-    final long msb;
+    private long msb;
 
     /**
      * Counter for the UUID LSB.
@@ -77,13 +77,13 @@ public class IncrementingUuidGenerator implements UuidGenerator {
 
     /**
      * Defines a new classloaderId for the class. This only affects instances
-     * created after the call (the instances created before the call keep their
-     * classloaderId). This method should be called to specify a classloaderId
-     * if you are using more than one class loader, and you want to guarantee a
-     * collision-free UUID generation (instead of the default random
-     * classloaderId which produces about 1% collision rate on the
-     * classloaderId, and thus can have UUID collision if the epoch-time,
-     * session counter and counter have the same values).
+     * created after the first call to {@link #generateId()} (the instances
+     * created before the call keep their classloaderId). This method should be
+     * called to specify a {@code classloaderId} if you are using more than one
+     * class loader, and you want to guarantee a collision-free UUID generation
+     * (instead of the default random classloaderId which produces about 1%
+     * collision rate on the classloaderId, and thus can have UUID collision if
+     * the epoch-time, session counter and counter have the same values).
      *
      * @param classloaderId the new classloaderId (only the least significant 12
      *                      bits are used)
@@ -94,6 +94,10 @@ public class IncrementingUuidGenerator implements UuidGenerator {
     }
 
     public IncrementingUuidGenerator() {
+
+    }
+
+    private long initializeMsb() {
         long sessionId = sessionCounter.incrementAndGet();
         if (sessionId == MAX_SESSION_ID) {
             throw new CucumberException(
@@ -103,7 +107,7 @@ public class IncrementingUuidGenerator implements UuidGenerator {
         }
         long epochTime = System.currentTimeMillis();
         // msb = epochTime | sessionId | version | classloaderId
-        msb = ((epochTime & MAX_EPOCH_TIME) << 24) | (sessionId << 16) | (8 << 12) | classloaderId;
+        return ((epochTime & MAX_EPOCH_TIME) << 24) | (sessionId << 16) | (8 << 12) | classloaderId;
     }
 
     /**
@@ -114,6 +118,11 @@ public class IncrementingUuidGenerator implements UuidGenerator {
      */
     @Override
     public UUID generateId() {
+        if (msb == 0) {
+            // Lazy init to avoid starting sessions when not used.
+            msb = initializeMsb();
+        }
+
         long counterValue = counter.incrementAndGet();
         if (counterValue == MAX_COUNTER_VALUE) {
             throw new CucumberException(

--- a/cucumber-core/src/test/java/io/cucumber/core/eventbus/IncrementingUuidGeneratorTest.java
+++ b/cucumber-core/src/test/java/io/cucumber/core/eventbus/IncrementingUuidGeneratorTest.java
@@ -1,8 +1,8 @@
 package io.cucumber.core.eventbus;
 
 import io.cucumber.core.exception.CucumberException;
-import org.hamcrest.Matchers;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.function.ThrowingSupplier;
 
 import java.io.IOException;
 import java.lang.reflect.Field;
@@ -21,6 +21,8 @@ import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
@@ -119,20 +121,33 @@ class IncrementingUuidGeneratorTest {
 
         // Then
         assertThat(cucumberException.getMessage(),
-            Matchers.containsString("Out of IncrementingUuidGenerator capacity"));
+            containsString("Out of IncrementingUuidGenerator capacity"));
     }
 
     @Test
     void version_overflow() {
         // Given
+        IncrementingUuidGenerator generator = new IncrementingUuidGenerator();
         IncrementingUuidGenerator.sessionCounter.set(IncrementingUuidGenerator.MAX_SESSION_ID - 1);
 
         // When
-        CucumberException cucumberException = assertThrows(CucumberException.class, IncrementingUuidGenerator::new);
+        CucumberException cucumberException = assertThrows(CucumberException.class, generator::generateId);
 
         // Then
         assertThat(cucumberException.getMessage(),
-            Matchers.containsString("Out of IncrementingUuidGenerator capacity"));
+            containsString("Out of IncrementingUuidGenerator capacity"));
+    }
+
+    @Test
+    void lazy_init() {
+        // Given
+        IncrementingUuidGenerator.sessionCounter.set(IncrementingUuidGenerator.MAX_SESSION_ID - 1);
+
+        // When
+        ThrowingSupplier<IncrementingUuidGenerator> instantiateGenerator = IncrementingUuidGenerator::new;
+
+        // Then
+        assertDoesNotThrow(instantiateGenerator);
     }
 
     private static void checkUuidProperties(List<UUID> uuids) {

